File: 多項式迴歸/多項式迴歸.py
#多項式迴歸
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt

data = pd.read_csv('Position_salaries.csv')
x = data.iloc[:, 1:2].values #需要是矩陣(不然後面會有問題)
y = data.iloc[:, 2].values

#只有十個數據，不需要分成訓練測試集
#迴歸不需要特徵縮放
#建立線性模型與多項式迴歸模型

#線性
from sklearn.linear_model import LinearRegression
model1 = LinearRegression()
model1.fit(x, y)

#多項式
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4) #n次多項式
x_poly = poly_reg.fit_transform(x) #常數項、一次項、二次項
model2 = LinearRegression()
model2.fit(x_poly, y)

#兩個迴歸模形圖
#線性
plt.scatter(x, y, color = 'red')
plt.plot(x, model1.predict(x), color = 'blue')
plt.title('Truth or  Bluff(linear regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

#多項式迴歸
x_grid = np.arange(min(x), max(x), 0.1)
x_grid = x_grid.reshape(len(x_grid), 1) #或者使用reshape(-1, 1)
plt.scatter(x, y, color = 'red')
plt.plot(x_grid, model2.predict(poly_reg.fit_transform(x_grid)), color = 'blue')
plt.title('Truth or  Bluff(Polynomial regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

File: 多項式迴歸/多項式迴歸.R
#多項式迴歸
#讀取資料
data = read.csv("position_salaries.csv")
data = data[, 2:3]

#不需要測試訓練集(數據資料太少)
#不需要特徵縮放(迴歸不需)
#線性模型
attach(data)
model1 = lm(Salary ~ Level, data)
summary(model1)

#多項式迴歸
data$level2 = data$Level^2 #平方項
data$level3 = data$Level^3 #三次方項
data$level4 = data$Level^4 #四次方項
model2 = lm(Salary ~ ., data)
summary(model2)

#畫圖
library(ggplot2)

#線性迴歸
ggplot() + 
  geom_point(aes (x = Level, y = Salary),
             color = 'red') + 
  geom_line(aes (x = Level, y = predict (model1,data)),
            color = 'blue') +
  xlab('Level') +
  ylab('Salary')

#多項式迴歸
ggplot() + 
  geom_point(aes (x = Level, y = Salary),
             color = 'red') + 
  geom_line(aes (x = Level, y = predict (model2,data)),
            color = 'blue') +
  xlab('Level') +
  ylab('Salary')

#預測線性模型結果
y_pred1 = predict(model1, data.frame(Level = 6.5))

#預測多項式模型結果
y_pred2 = predict(model2, data.frame(Level = 6.5,
                                     level2 = 6.5 ^ 2,
                                     level3 = 6.5 ^ 3,
                                     level4 = 6.5 ^ 4))
